src/object.rs

```diff
@@ -20,9 +20,9 @@ where
     ///
     /// It is okay for existing readers to still refer to Self.
     ///   
-    unsafe fn retire(self: *mut Self, deleter: &'static dyn Deleter) -> usize {
-        let ptr = self as *mut (dyn Reclaim + 'domain);
-        unsafe { (&*self).domain().retire(ptr, deleter) }
+    unsafe fn retire(this: *mut Self, deleter: &'static dyn Deleter) -> usize {
```

Comment:
Can you leave a `TODO` comment here mentioning how this should be `self` once `arbitrary_self_types` stabilized along with a link to the [tracking issue](https://github.com/rust-lang/rust/issues/44874)?